hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AFileSystem.java

```diff
@@ -2603,6 +2607,30 @@ S3AFileStatus innerGetFileStatus(final Path f,
     entryPoint(INVOCATION_GET_FILE_STATUS);
     checkNotClosed();
     final Path path = qualify(f);
+    return resolveFileStatus(path, needEmptyDirectoryFlag, false);
+  }
+
+
+  /**
+   * Get the status of a file or directory, first through S3Guard and then
+   * through S3.
+   * The S3 probes can leave 404 responses in the S3 load balancers; if
+   * a check is only needed for a directory, declaring this saves time and
+   * avoids creating one for the object.
+   * When only probing for directories, if an entry for a file is found in
+   * S3Guard it is returned, but checks for updated values are skipped.
+   * @param path fully qualified path
+   * @param needEmptyDirectoryFlag if true, implementation will calculate
+   *        a TRUE or FALSE value for {@link S3AFileStatus#isEmptyDirectory()}
+   * @param onlyProbeForDirectory skip the simple object probe
+   * @return a S3AFileStatus object
+   * @throws FileNotFoundException when the path does not exist
+   * @throws IOException on other problems.
+   */
+  private S3AFileStatus resolveFileStatus(final Path path,
```

Comment:
Do you plan to refactor this to a class - like S3AFileStatusResolver? What's the reasoning behind factoring this out from innerGetFileStatus? This leaves that method just a call for this method - plus a few other things, but basically, it's just that.

Reply:
I think if we plan to refactor something out from S3AFs it would be good to start now :)

Reply:
I'm avoiding that now for what is really a bug fix, especially as we need a good layering design first. Yes, the s3* calls are clearly at the bottom, but unwinding them will take more effort than I want to. I've stuck the new probe enum into the .impl package though.